cpp/serve/sampler/cpu_sampler.cc

```diff
@@ -391,6 +393,20 @@ class CPUSampler : public SamplerObj {
             sample_results[i].push_back(sample_result);
             break;
           }
+          // if cur_token_idx == verify_end - verify_start - 1
+          // all draft tokens are accepted
+          // we sample a new token
+          if (cur_token_idx == verify_end - verify_start - 1) {
+            SampleResult sample_result;
+            // sample a new token from the original distribution
+            sample_result.sampled_token_id = SampleTopPFromProb(
+                probs_host, verify_start + cur_token_idx,
+                generation_cfg[i]->temperature < eps_ ? 0.0 : generation_cfg[i]->top_p,
+                rngs[i]->GetRandomNumber());
+            sample_result.top_prob_tokens = ComputeTopProbs(
+                probs_host, verify_start + cur_token_idx, generation_cfg[i]->top_logprobs);
+            sample_results[i].push_back(sample_result);
+          }
```

Comment:
I'm trying to understand the logic here. Does it mean that the last prediction is always ignored?

I'm also wondering that whether the `BatchVerifyDraftTokens` after updates in this PR is compatible with the speculative decoding prior to this PR?

Reply:
We verify the last token. All the predictions are used. The `-1` in `verify_end - verify_start - 1` is because `verify_end - verify_start` is the length of `[last_committed_token, 1st_predict_token, 2nd_predict_token, ...., last_predict_token]`, we don't need to verify `last_committed_token`. The condition `if (cur_token_idx == verify_end - verify_start - 1)` here means that all draft tokens are accepted, we sample a new token according to the base model logits.